homeassistant/components/binary_sensor/w800rf32.py

```diff
@@ -0,0 +1,179 @@
+"""
+Support for w800rf32 binary sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/binary_sensor.w800rf32/
+
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASSES_SCHEMA, PLATFORM_SCHEMA, BinarySensorDevice)
+from homeassistant.components.w800rf32 import (
+    DOMAIN, CONF_FIRE_EVENT, CONF_OFF_DELAY)
+from homeassistant.const import (
+    ATTR_ENTITY_ID, ATTR_STATE,
+    CONF_DEVICE_CLASS, CONF_NAME,
+    ATTR_NAME, CONF_DEVICES)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers import event as evt
+from homeassistant.util import dt as dt_util
+from homeassistant.helpers.dispatcher import (dispatcher_connect)
+
+_LOGGER = logging.getLogger(__name__)
+
+DEPENDENCIES = ['w800rf32']
+DEVICE_TYPE = 'W800rf32BinarySensor'
+
+ATTR_FIRE_EVENT = 'fire_event'
+EVENT_BUTTON_PRESSED = 'button_pressed'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICES, default={}): {
+        cv.string: vol.Schema({
+            vol.Optional(CONF_NAME): cv.string,
+            vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
+            vol.Optional(CONF_FIRE_EVENT, default=False): cv.boolean,
+            vol.Optional(CONF_OFF_DELAY):
+            vol.Any(cv.time_period, cv.positive_timedelta)
+        })
+    },
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Binary Sensor platform to w800rf32."""
+    import W800rf32 as w800rf32mod
+    binary_sensors = []
+    # device_id --> "c1 or a3" X10 device. entity (type dictionary)
+    # --> name, device_class etc
+    for device_id, entity in config[CONF_DEVICES].items():
+
+        if device_id in hass.data[DOMAIN]['entities']:
+            continue
+
+        _LOGGER.debug("Add %s w800rf32.binary_sensor (class %s)",
+                      entity[ATTR_NAME], entity.get(CONF_DEVICE_CLASS))
+
+        device = W800rf32BinarySensor(
+            device_id, entity.get(CONF_NAME), entity.get(CONF_DEVICE_CLASS),
+            entity[CONF_FIRE_EVENT], entity.get(CONF_OFF_DELAY))
+
+        hass.data[DOMAIN]['entities'][device_id] = device
```

Comment:
We shouldn't and don't need to store the entities in this shared dictionary. Move the entity event callback update function inside the entity class as a method. Then we can signal the entities that match the device id directly without storing the entities in a list.

Connect the event callback in `async_added_to_hass` entity coroutine method.